src/client/Client.cc

```diff
@@ -5309,24 +5309,48 @@ void Client::handle_caps(const MConstRef<MClientCaps>& m)
 
   got_mds_push(session.get());
 
+  bool do_cap_release = false;
   Inode *in;
   vinodeno_t vino(m->get_ino(), CEPH_NOSNAP);
   if (auto it = inode_map.find(vino); it != inode_map.end()) {
     in = it->second;
+
+    /* MDS maybe waiting for cap release with increased seq */
+    switch (m->get_op()) {
+      case CEPH_CAP_OP_REVOKE:
+      case CEPH_CAP_OP_GRANT:
+        if (!in->caps.count(mds)) {
+	  do_cap_release = true;
+          ldout(cct, 5) << __func__ << " vino " << vino << " don't have cap "
+                        << m->get_cap_id() << " op " << m->get_op()
+                        << ", immediately releasing" << dendl;
+	}
+    }
   } else {
-    if (m->get_op() == CEPH_CAP_OP_IMPORT) {
-      ldout(cct, 5) << __func__ << " don't have vino " << vino << " on IMPORT, immediately releasing" << dendl;
-      session->enqueue_cap_release(
-        m->get_ino(),
-        m->get_cap_id(),
-        m->get_seq(),
-        m->get_mseq(),
-        cap_epoch_barrier);
-    } else {
-      ldout(cct, 5) << __func__ << " don't have vino " << vino << ", dropping" << dendl;
+    /* MDS maybe waiting for cap release with increased seq */
+    switch (m->get_op()) {
+      case CEPH_CAP_OP_IMPORT:
+      case CEPH_CAP_OP_REVOKE:
+      case CEPH_CAP_OP_GRANT:
+	do_cap_release = true;
+        ldout(cct, 5) << __func__ << " don't have vino " << vino << " op "
+                      << m->get_op() << ", immediately releasing" << dendl;
+	break;
+      default:
+        ldout(cct, 5) << __func__ << " don't have vino " << vino << ", dropping" << dendl;
+	return;
     }
+  }
+
+  // In case the mds is waiting on e.g. a revocation
+  if (do_cap_release) {
+    session->enqueue_cap_release(
+      m->get_ino(),
+      m->get_cap_id(),
+      m->get_seq(),
```

Comment:
This is problematic (change also in the kernel driver), causing:

https://tracker.ceph.com/issues/64977

It should be the last `issue_seq` but at this point the client does not have a valid `issue_seq` because the cap is already removed. Furthermore, the MDS does not share its view of the `issue_seq` associated with the grant/revoke; sadly only clients actually set the `issue_seq` in an `MClientCaps` message header.

For the `CEPH_CAP_OP_IMPORT` case, it was okay to use `m->get_seq()` because that's implicitly the `issue_seq` as well but this is not okay for the `CEPH_CAP_OP_GRANT` or `CEPH_CAP_OP_REVOKE` cases.

I'm confused by how the original issue occurs with the stated sequence in this PR's description:

https://tracker.ceph.com/issues/57244

specifically the second step on the MDS side "Receives release req and...". The MDS does not increase `issue_seq` when sending revokes. The client's release should be processed successfully. This is an expected sequence of events which was accounted for as long ago as b180b6ce1981b6d23b68273266a0ab6594f40824. I think there was something else going on here that should be investigated.

With that said, we're now obligated to support this new scenario where the client sends a cap release with `issue_seq == (last revoke/grant) seq`. This only happens if the client is basically saying: "I don't acknowledge that I have this cap; please release it.". The MDS should be able to handle this gracefully and that's basically done in #56828.

I'm in the middle of some cleanup relating to `issue_seq` but wanted to leave this note here for anyone looking at this in the future. Cleanup tracked in https://tracker.ceph.com/issues/66704.

Reply:
cc @lxbsz @vshankar @kotreshhr

Reply:
> For the `CEPH_CAP_OP_IMPORT` case, it was okay to use `m->get_seq()` because that's implicitly the `issue_seq` as well but this is not okay for the `CEPH_CAP_OP_GRANT` or `CEPH_CAP_OP_REVOKE` cases.

Agreed. For the wider audience, the last_sent/last_issue/mseq has deep history to solve particularly this release/revoke race which unfortunately broke with this commit. 

> specifically the second step on the MDS side "Receives release req and...". The MDS does not increase `issue_seq` when sending revokes. The client's release should be processed successfully. This is an expected sequence of events which was accounted for as long ago as [b180b6c](https://github.com/ceph/ceph/commit/b180b6ce1981b6d23b68273266a0ab6594f40824). I think there was something else going on here that should be investigated.

... and this wasn't caught with fs suite runs which makes me worried a lot. Yes, it's a concurrent release/revoke case, but really we aren't testing to the extent where such cases are covered (esp. since we have now seen a couple of reports for this bug, one without much scale involved).

Reply:
> This is problematic (change also in the kernel driver), causing:
> 
> https://tracker.ceph.com/issues/64977
> 
> It should be the last `issue_seq` but at this point the client does not have a valid `issue_seq` because the cap is already removed. 

How could this happen is what I eagerly want to figure out, after that I think we can resolve all the confusion about the revoke stuck issues.

> Furthermore, the MDS does not share its view of the `issue_seq` associated with the grant/revoke; sadly only clients actually set the `issue_seq` in an `MClientCaps` message header.
> 

The `issue_seq` in client will be updated only together with the `client request reply` msgs. 


> For the `CEPH_CAP_OP_IMPORT` case, it was okay to use `m->get_seq()` because that's implicitly the `issue_seq` as well but this is not okay for the `CEPH_CAP_OP_GRANT` or `CEPH_CAP_OP_REVOKE` cases.
> 
> I'm confused by how the original issue occurs with the stated sequence in this PR's description:
> 
> https://tracker.ceph.com/issues/57244
> 
> specifically the second step on the MDS side "Receives release req and...". The MDS does not increase `issue_seq` when sending revokes. The client's release should be processed successfully. This is an expected sequence of events which was accounted for as long ago as [b180b6c](https://github.com/ceph/ceph/commit/b180b6ce1981b6d23b68273266a0ab6594f40824). I think there was something else going on here that should be investigated.
> 
> With that said, we're now obligated to support this new scenario where the client sends a cap release with `issue_seq == (last revoke/grant) seq`. This only happens if the client is basically saying: "I don't acknowledge that I have this cap; please release it.". The MDS should be able to handle this gracefully and that's basically done in #56828.
> 
> I'm in the middle of some cleanup relating to `issue_seq` but wanted to leave this note here for anyone looking at this in the future. Cleanup tracked in https://tracker.ceph.com/issues/66704.